### PR TITLE
addition of optional max FFD spline order and optional exclude wrt parameter

### PR DESCRIPTION
**Pull request by anilyil**
## Purpose

This PR introduces two small changes:
1. When creating a DVGeo object using FFDs, we can now specify the maximum order of the splines used in the FFD formulation with the `kmax` parameter, which was hard-coded to 4 previously. A lower order is automatically selected based on the dimensions of the FFD. That behavior is unchanged.
2. Added an optional argument `exclude_wrt` to `addConstraintsPyOpt` in DVCon. This is useful if we want to omit some DVs from geometric constraints manually. 

I am not testing these yet, but I can add tests based on what people suggest. We can also modify existing tests to check for these.

## Type of change
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (non-backwards-compatible fix or feature)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Documentation update
- [ ] Maintenance update
- [ ] Other (please describe)

## Checklist

- [ ] I have run `flake8` and `black` to make sure the code adheres to PEP-8 and is consistently formatted
- [ ] I have run unit and regression tests which pass locally with my changes
- [ ] I have added new tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation

### CODE REVIEW

**Comment by marcomangano** [comment]
The additions look good, thanks for fixing this. Two comments:
- The test failures are happening because of the array order is not the same on different images. We have seen this before but I don't remember if we concluded anything about it. Maybe this is a good opportunity to take care of it?
- I would add just a couple of tests, one in `test_DVconstraints` for the `except_wrt` option and one in `test_DVGeometry` for the spline ordering. I would take one of the existing cases and replicate it with the new options (we should switch to `parameterize` at some point...). For the first case it should be pretty easy, in the second one we need to find a case where the different spline order is creating a noticeable difference, hopefully we can re-use existing FFD files

File: pygeo/pyBlock.py [hunk 1]
```diff
@@ -60,7 +60,7 @@ def __init__(self, initType, fileName=None, FFD=False, symmPlane=None, **kwargs)
     #                     Initialization Types
     # ----------------------------------------------------------------------
 
-    def _readPlot3D(self, fileName, order="f", FFD=False, symmTol=0.001):
+    def _readPlot3D(self, fileName, order="f", FFD=False, symmTol=0.001, kmax=4):
```
Comment (nwu63) [inline pygeo/pyBlock.py:63]: I'm guessing this is one of the `kwargs` that gets passed by the `DVGeometry` constructor call? In that case, please document this option both here, and in the `DVGeometry` init call.

File: pygeo/constraints/baseConstraint.py [hunk 2]
```diff
@@ -58,13 +58,23 @@ def getVarNames(self):
         """
         return self.DVGeo.getVarNames()
 
-    def addConstraintsPyOpt(self, optProb):
+    def addConstraintsPyOpt(self, optProb, exclude_wrt=None):
```
Comment (nwu63) [inline pygeo/constraints/baseConstraint.py:61]: Also need docstrings

**Review by nwu63 — CHANGES_REQUESTED** [review]
I was trying to decide if having `exclude_wrt` or `wrt` would be better, but I think the exclude makes more sense as in general we want to include everything. See comments below in addition to what Marco said

File: pygeo/pyBlock.py (hunk 1 above)
Comment (nwu63) [reply pygeo/pyBlock.py:63]: I guess this is actually passed through the `pyBlock` constructor as well. I don't like undocumented `kwargs` so we should try to do a better job here.

**Comment by joanibal** [comment]
I suggest making the order in each direction a parameter. 
If it's alright with you I'll update the PR with this feature, which I have already implemented on my dev fork.

**Comment by joanibal** [comment]
I'm good with the PR after Neil's comments have been addressed.

**Comment by joanibal** [comment]
Anil and I talked over slack about removing the option to have different orders in different directions.

**Review by joanibal — DISMISSED** [review]

File: pygeo/pyBlock.py (hunk 1 above)
Comment (anilyil) [reply pygeo/pyBlock.py:72]: tried to clarify how this is done a bit by exposing the kmax in the dvgeo init and the pyblock init. please take a look and let me know if its okay now

File: pygeo/constraints/baseConstraint.py (hunk 2 above)
Comment (anilyil) [reply pygeo/constraints/baseConstraint.py:61]: added. let me know if you want any other changes here

**Comment by anilyil** [comment]
I addressed the comments on docstrings. will work on adding the tests now

**Comment by anilyil** [comment]
I added a test on the custom FFD spline order. It fails when the order is not set correctly. I am not sure how I can test the `exclude_wrt` option. Do we test the `addConstraintsPyOpt` methods? If not testing the `exclude_wrt` stuff is okay, this one is good for another review.

**Review by joanibal — APPROVED** [review]

**Review by nwu63 — APPROVED** [review]
LGTM

**Comment by nwu63** [comment]
> I added a test on the custom FFD spline order. It fails when the order is not set correctly. I am not sure how I can test the `exclude_wrt` option. Do we test the `addConstraintsPyOpt` methods? If not testing the `exclude_wrt` stuff is okay, this one is good for another review.

I'm okay with not testing the `exclude_wrt` option for the time being.

**Pull request merged**